File: src/prompt.rs
use crate::session::{CommentStatus, CopyBatch, Session};
use chrono::Utc;
use sha2::{Digest, Sha256};
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CopyScope {
    Uncopied,
    AllVisible,
}
pub fn build_prompt(session: &Session, scope: CopyScope) -> Option<(String, Vec<String>)> {
    let mut cs: Vec<_> = session
        .visible_comments()
        .into_iter()
        .filter(|c| scope == CopyScope::AllVisible || c.status == CommentStatus::Draft)
        .collect();
    if cs.is_empty() {
        return None;
    }
    cs.sort_by_key(|c| {
        (
            c.anchor.repo.clone(),
            c.anchor.file.clone(),
            c.anchor.old_line.or(c.anchor.new_line).unwrap_or(0),
        )
    });
    let mut out=String::from("Please address these local Hawk review comments. Keep changes focused and reply with a summary.\n");
    let (mut repo, mut file) = (String::new(), String::new());
    let mut ids = Vec::new();
    for c in cs {
        if c.anchor.repo != repo {
            repo = c.anchor.repo.clone();
            out.push_str(&format!("\n## Repo `{repo}`\n"));
            file.clear();
        }
        if c.anchor.file != file {
            file = c.anchor.file.clone();
            out.push_str(&format!("\n### `{file}`\n"));
        }
        ids.push(c.id.clone());
        let line = match c.anchor.side {
            crate::diff::Side::Old => {
                format!("old line {} (removed)", c.anchor.old_line.unwrap_or(0))
            }
            crate::diff::Side::New => {
                format!("new line {} (added)", c.anchor.new_line.unwrap_or(0))
            }
            crate::diff::Side::Both => format!(
                "line {} (context)",
                c.anchor.new_line.or(c.anchor.old_line).unwrap_or(0)
            ),
        };
        out.push_str(&format!(
            "- {} ({})\n  Hunk: `{}`\n  Context: `{}`\n  Comment:\n{}\n",
            file,
            line,
            c.anchor.hunk_header,
            c.anchor.line_text,
            format_comment_body(&c.body)
        ));
    }
    Some((out, ids))
}
fn format_comment_body(body: &str) -> String {
    body.trim_end()
        .lines()
        .map(|line| format!("    {line}"))
        .collect::<Vec<_>>()
        .join("\n")
}

pub fn mark_copied(session: &mut Session, ids: &[String], scope: &str, prompt: &str) {
    for c in session.comments.values_mut() {
        if ids.contains(&c.id) && c.status == CommentStatus::Draft {
            c.status = CommentStatus::Copied
        }
    }
    let mut h = Sha256::new();
    h.update(prompt);
    session.batches.push(CopyBatch {
        id: format!("batch-{}", session.batches.len() + 1),
        timestamp: Utc::now(),
        scope: scope.into(),
        comment_ids: ids.to_vec(),
        prompt_hash: format!("{:x}", h.finalize()),
    });
}
#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        diff::{DiffLine, LineKind},
        session::{LineAnchor, Session},
    };
    #[test]
    fn multiline_comments_render_as_an_indented_block_without_trailing_blank_lines() {
        let mut s = Session::default();
        let l = DiffLine {
            kind: LineKind::Add,
            old_lineno: None,
            new_lineno: Some(3),
            text: "new".into(),
        };
        s.upsert_comment(
            LineAnchor::new("repo", "new.rs", "@@ -1 +1", &l),
            "abcd\nefg\n\n".into(),
        );

        let (p, _) = build_prompt(&s, CopyScope::Uncopied).unwrap();

        assert!(p.contains("  Comment:\n    abcd\n    efg\n"));
        assert!(!p.contains("efg\n    \n"));
    }

    #[test]
    fn prompt_group_and_mark() {
        let mut s = Session::default();
        let l = DiffLine {
            kind: LineKind::Remove,
            old_lineno: Some(2),
            new_lineno: None,
            text: "bad".into(),
        };
        s.upsert_comment(
            LineAnchor::new("repo", "old.rs", "@@ -2", &l),
            "why?".into(),
        );
        let (p, ids) = build_prompt(&s, CopyScope::Uncopied).unwrap();
        assert!(p.contains("old line 2 (removed)"));
        mark_copied(&mut s, &ids, "uncopied", &p);
        assert_eq!(s.batches.len(), 1);
        assert!(build_prompt(&s, CopyScope::Uncopied).is_none());
    }
}

File: src-ts/core/prompt.ts
import { createHash } from "node:crypto"
import type { Comment, Session } from "./model.js"
import { visibleComments } from "./session.js"

export type CopyScope = "uncopied" | "all-visible"

export function buildPrompt(session: Session, scope: CopyScope): { prompt: string; ids: string[] } | undefined {
  const comments = visibleComments(session)
    .filter((c) => scope === "all-visible" || c.status === "draft")
    .sort((a, b) => `${a.anchor.repo}\0${a.anchor.file}\0${a.anchor.oldLine ?? a.anchor.newLine ?? 0}`.localeCompare(`${b.anchor.repo}\0${b.anchor.file}\0${b.anchor.oldLine ?? b.anchor.newLine ?? 0}`))
  if (comments.length === 0) return undefined

  let out = "Please address these local Hawk review comments. Keep changes focused and reply with a summary.\n"
  const ids: string[] = []
  let repo = ""
  let file = ""
  for (const comment of comments) {
    if (comment.anchor.repo !== repo) {
      repo = comment.anchor.repo
      out += `\n## Repo \`${repo}\`\n`
      file = ""
    }
    if (comment.anchor.file !== file) {
      file = comment.anchor.file
      out += `\n### \`${file}\`\n`
    }
    ids.push(comment.id)
    out += formatComment(comment)
  }
  return { prompt: out, ids }
}

export function markCopied(session: Session, ids: string[], scope: string, prompt: string): Session {
  const comments = { ...session.comments }
  for (const [key, comment] of Object.entries(comments)) {
    if (ids.includes(comment.id) && comment.status === "draft") comments[key] = { ...comment, status: "copied" }
  }
  return {
    ...session,
    comments,
    batches: [...session.batches, { id: `batch-${session.batches.length + 1}`, timestamp: new Date().toISOString(), scope, commentIds: ids, promptHash: createHash("sha256").update(prompt).digest("hex") }],
  }
}

function formatComment(comment: Comment): string {
  const line = comment.anchor.side === "old"
    ? `old line ${comment.anchor.oldLine ?? 0} (removed)`
    : comment.anchor.side === "new"
      ? `new line ${comment.anchor.newLine ?? 0} (added)`
      : `line ${comment.anchor.newLine ?? comment.anchor.oldLine ?? 0} (context)`
  return `- ${comment.anchor.file} (${line})\n  Hunk: \`${comment.anchor.hunkHeader}\`\n  Context: \`${comment.anchor.lineText}\`\n  Comment:\n${formatBody(comment.body)}\n`
}

function formatBody(body: string): string {
  return body.trimEnd().split(/\r?\n/).map((line) => `    ${line}`).join("\n")
}
